sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/JavaUDF.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, CodeGenerator, ExprCode}
+import org.apache.spark.sql.types.DataType
+
+/*
+ Discussion: the ctor arg `function` need to support both scala and java.
+  so I think the style `(Any, Any, ...) => Any` is easier to use
+  Java function can be passed in using the value `javaFunction.call(_:Any, ...)`
+ */
+private[sql] case class JavaUDF (
```

Comment:
Doesn't this looks as same as current `ScalaUDF`? The main difference is?

Reply:
yes, almost the same for now, but the PR will change later after more discussion. (including interface and implementation), no worry!